### PR TITLE
Add ability to configure which hooks to run per scan

### LINKED ISSUE #728: Configure Hooks to run per scan
## ➹ New Feature implementation request

It would be extremely helpful if you can decide on a per-scan basis which hooks to run.

### Example use case

Installed hooks: Cascading Scans, DefectDojo persistence provider

Started scan:

```yaml
apiVersion: "execution.securecodebox.io/v1"
kind: Scan
metadata:
  name: "nmap-open-ports"
spec:
  scanType: "nmap"
  parameters:
    - "-p-"
    # Against Host
    - "example.com"
```

The initial scan would create this finding:

```json
 {
    "name":"Open Port: 443 (http)",
    "category":"Open Port",
    "attributes":{
       "port":443,
       "state":"open",
       "service":"http",
       "serviceProduct":"nginx",
       "serviceVersion":null,
       "tunnel":"ssl"
    },
 },
```

Cascading Rule:

```yaml
apiVersion: "cascading.securecodebox.io/v1"
kind: CascadingRule
metadata:
  name: "nmap-service-detection"
spec:
  matches:
    anyOf:
      - category: "Open Port"
        attributes:
          state: open
  scanSpec:
    scanType: "nmap"
    parameters:
      - "-p{{attributes.port}}"
      - "-sV"
      - "--service-all"
      # Against Host
      - "{{$.hostOrIP}}"
```

When triggered, the Cascading Scan would create the following finding with **more service information**:

```json
 {
    "name":"Open Port: 443 (http)",
    "category":"Open Port",
    "attributes":{
       "port":443,
       "state":"open",
       "service":"http",
       "serviceProduct":"nginx",
       "serviceVersion":"1.20.1",
       "tunnel":"ssl"
    },
 },
```

In this case, I would like to **not import** the initial scan results into DefectDojo but still run the Cascading Scan hook.

### Describe the solution you'd like
Once #695 is merged, we could use Hook Priorities to solve this problem. One could deploy Cascading Scans with a priority of `1` and DefectDojo with `0`. Then on a per-scan basis one may define what hook ranges to execute.

```yaml
apiVersion: "execution.securecodebox.io/v1"
kind: Scan
metadata:
  name: "nmap-open-ports"
spec:
  hookRanges:
    - "1"
  scanType: "nmap"
  parameters:
    - "-p-"
    # Against Host
    - "example.com"
```

`hookRanges` would be a list of ranges to execute. A range could have the following formats:

* `0-1`: Execute hooks with priority between `0` and `1` (inclusive)
* `0`: Execute hooks with priority `0`.

The `orderedHookStatusses` status field would still include the skipped hooks but marks them with state `Skipped`.

### Describe alternatives you've considered
Proposed workarounds involved setting up multiple namespaces with different hooks installed and then running the scan in that namespace.

### PR DESCRIPTION
**Pull request by EndPositive**
## Description
This PR, if applied, closes #728, mainly by adding the field `hookSelector` in the scan spec. Besides, this PR adds the cascading scans field `inheritHookSelector` (default `false`) for more control over the `hookSelector` field in cascading scans.

Related documentation PR secureCodeBox/documentation#152

### Example usage

To select all hooks, leave the `hookSelector` field undefined or specify:
```yaml
hookSelector: {}
```

To select specific hooks:

```yaml
hookSelector:
  matchExpressions:
   - key: app.kubernetes.io/instance
     operator: In
     values: [ "defectdojo", "update-field" ]
```

To ignore certain hooks:
```yaml
hookSelector:
  matchExpressions:
  -  key: app.kubernetes.io/instance
     operator: NotIn
     values: [ "defectdojo" ]
```

Available operators are:
`In`, `NotIn`, `Exists`, `DoesNotExist`

You may also use `matchLabels` for exact matches on labels:
```yaml
hookSelector:
  matchLabels:
    app.kubernetes.io/instance: "defectdojo"
```

`matchLabels` and `matchExpressions` may be used simultaneously.

### Default Hook labels
To select a certain hook deployment, use the label `app.kubernetes.io/instance` (e.g. `update-severity` for the `update-field` hook). To select a certain hook type, use the label `app.kubernetes.io/name` (e.g. `update-field`).

Extra hook labels may be added by deploying the hook with the `hook.labels` field set in `values.yaml`.

### Usage in combination with Cascading Scans

Besides existing fields (e.g. `inheritVolumes`, `inheritLabels`, etc.), the field `inheritHookSelector` is now available with the same behavior as the other existing fields.

**Note:** to use cascading scans in combination with `hookSelector`, ensure that you also select the cascading scans hook. You can use the existing label `securecodebox.io/internal` to select core features like cascading scans. You should manually ensure that your other selectors override this selector in behavior.

```yaml
hookSelector:
  matchLabels:
    securecodebox.io/internal: "true"
```

### Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ ] Make sure `npm test` runs for the whole project.
* [x] Make codeclimate checks happy

### CODE REVIEW

File: hooks/cascading-scans/hook/hook.test.js [hunk 1]
```diff
@@ -1595,6 +1621,18 @@ test("should purge cascaded scan spec from parent scan", () => {
...
+    Object {
+      "key": "securecodebox.io/name",
+      "operator": "In",
+      "values": Array [
+        "cascading-scans",
+      ],
+    },
+  ]
+  `)
+  expect(secondCascadedScan.spec.hookSelector.matchLabels).toMatchInlineSnapshot(`Object {}`)
```
[the remark this answers is not in the document]
Comment (EndPositive) [reply hooks/cascading-scans/hook/hook.test.js:1635]: @malexmave currently, the Cascading Scans hook does purging based on the current scan spec and the matched cascading rule. If an entry is overridden with the cascading rule, the Cascading Scans hook has no way to retrieve the original entry.

> Is there a reason why the code cannot behave like that? ...

Yeah that _was_ the reason. Over the weekend, I've been thinking of a few ways to solve the inheritance problem we're having here.

## Option 1 - Retrieve root scan to create cascaded scan
Using `ownerReferences`, the Cascading Scans hook could recurse all the way back to the initial/root scan. Then merge the entries from the matched Cascading Rule.

## Option 2 - Add "Engagement" CRD for global entries
We could all a global CRD where we define global entries that should be inherited to each scan within that engagement. Cascading Scans hook retains same behaviour as Option 1 but now retrieving the Engagement resource.

## Option 3 - Specify inheritance per entry
For ultimate customization, users should be able to configure inheritance per entry. We could update our CRD's such that each `env`, `initContainer`, etc. contains an extra field `persist` which controls whether that entry should be passed on to cascaded scans. Kubernetes map types (e.g. labels) would be converted into structs.

All of the above described solutions would the issues you're describing.

In any case, I think the mentioned issues are present in the current state of secureCodeBox and are only applied additionally to this PR, but not introduced in this PR. Personally, I'm really fine with documenting side-effects of inheritance (Secura's use-case is not affected by the described issues). I think we should save the intuitiveness and behavioral issue of inheritance for future work, possibly using ADRs to discuss solutions described above.

[the remark this answers is not in the document]
Comment (malexmave) [reply hooks/cascading-scans/hook/hook.test.js:1635]: Okay, thank you for your thoughts and clarifications. I agree that this is not in scope for this pull request, I just wanted to understand what exactly was going on here. If we add this behavior to the docs for the new feature, I am fine with leaving it as-is for the moment and pulling any discussions about how to address this into a separate issue.

**Review by malexmave — APPROVED** [review]
Thank you again for your work on this. I think that this state is ready to merge. For documentation, I think we will need doc changes in the following places:
- Explanations of the CRDs (API reference)
- Explanations of how a scan is created (adding the selectors)
- Explanations for the different hooks (adding labels)
- Explanations for contributing a new hook (updating the examples given in the docs for values.yaml and the template)
- Explanations of cascading scans (new inheritance fields, behavior as discussed)
- Ideally, we could add another article in the "How to" section on "Select which hooks to run", which explains the whole process in an end-to-end way (installing hooks with custom labels, selecting which labels to run, ...).

Since #695 will require significant documentation changes in very similar areas as well, it may be easier to combine the documentation updates for both features. In that case, you could also make a more general how-to on "controlling hook execution" which covers the features introduced in both PRs together.

But this can be done in a separate PR (or two, if it covers both repos).

@J12934 is there an easy way to get the CI to run on this, or do we merge through the warning (I don't have the necessary rights for this at the moment). Also, at the moment CodeClimate [seems a bit miffed](https://codeclimate.com/github/secureCodeBox/secureCodeBox/pull/757) about the new code. I don't really see an easier way to implement the part it complains about though.

File: hooks/cascading-scans/hook/hook.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (malexmave) [reply hooks/cascading-scans/hook/hook.test.js:1635]: I have pulled this into its own issue: https://github.com/secureCodeBox/secureCodeBox/issues/789. Resolving the conversation here.

**Comment by EndPositive** [comment]
@malexmave see the new documentation PR  secureCodeBox/documentation#152. Feel free to add/correct wherever necessary. Did you want to make a new blogpost too?

**Comment by malexmave** [comment]
Thanks! I'll take a look at the PR later today or tomorrow at the latest.

For the blog post: I understood that you aren't eager to write the blogpost yourself? Because if you want, you can definitely do so (they are your features, after all :). If you don't want to write the post, I will see if I can find time to write one, but I don't consider it an absolute requirement to have a blog post, especially if we have a good documentation article (I see that you wrote one, I just haven't read it in detail yet).

In general: I consider this PR ready to merge, I'm just holding off on pressing the button until I've had time to also read the documentation PR, in case anything comes up while going through it.

**Comment by EndPositive** [comment]
> I understood that you aren't eager to write the blogpost yourself?

Indeed. I'm personally not a big fan of writing blog-posts (or documentation in general :grimacing: ), and we don't really have the capacity to work on these, also considering their value for Secura. Though, feel free to write any blog posts about anything we contribute, especially if you're so closely involved in the review process.

**Review by J12934 — APPROVED** [review]

**Pull request merged**